runtime/src/main/scala/com/armanbilge/vilcacora/runtime/Interpreter.scala

```
@@ -0,0 +1,305 @@
+    memoryResource(model).use { memoryMap =>
+      for {
+        _ <- copyInputsToMemory(inputs, model, memoryMap)
+
+        // The core execution loop. It executes each operation and then yields control
+        // to the Cats Effect runtime to maintain application responsiveness.
+        // This is the desired structure.
+        _ <- model.operations.traverse_ { op =>
+          executeOperation(op, memoryMap, model) <* IO.cede
+        }
+
+        results <- copyOutputsFromMemory(model, memoryMap)
+      } yield results
+    }
+
+  /** A `Resource` that manages the lifecycle of all tensor memory for the graph. */
+  private def memoryResource(model: ModelIR): Resource[IO, MemoryMap] = {
+    val allocAll = IO {
+      model.allocations.map { case (name, allocation) =>
+        val totalBytes = (allocation.shape.product * allocation.dataType.sizeInBytes).toUSize
+        val ptr = stdlib.malloc(totalBytes)
+        if (ptr == null) throw new OutOfMemoryError(s"Failed to allocate tensor '$name'")
+        allocation.initialData.foreach(data =>
+          memcpy(ptr, data.at(0).asInstanceOf[Ptr[Byte]], data.length.toUSize),
+        )
+        name -> ptr
+      }
+    }
+    Resource.make(allocAll)(memoryMap => IO(memoryMap.values.foreach(stdlib.free))).map(_.toMap)
+  }
+
+  /** Executes a single operation by dispatching to the appropriate handler. */
+  private def executeOperation(op: Operation, memory: MemoryMap, model: ModelIR): IO[Unit] =
+    op match {
+      case op: Operation.SVMClassifier => handleSvmClassifier(op, memory, model)
+      case op: Operation.Add => handleElementWise(op, memory, model)(_ + _)
+      case op: Operation.Mul => handleElementWise(op, memory, model)(_ * _)
+      case op: Operation.Cast => handleCast(op, memory, model)
+      case other =>
+        IO.raiseError(
+          new NotImplementedError(s"Operation not implemented: ${other.getClass.getSimpleName}"),
+        )
```

Similar idea here, we can actually raise this error outside of the `IO` so that it surfaces earlier.